Feature: add prompt load from layer

ElliottLester opened the pull request:
Scope:
We want to restore the state of SDImgPageBase from an already generated layer.

Requests for comment:
- Is this a reasonable design
- Is it worth making an "info" class for holding generation information but no the base64 encoded images. -- this allows to/from json
-- central list of properties
- I don't like the way update_prompt_params works at the moment.
- Should we update the model I don't because the user can't see it from the prompt screen. (expand info bar maybe?)

ElliottLester committed:
Feature add prompt load from layer

Scope:
We want to restore the state of SDImgPageBase from an already generated layer.
To do this we create a class to store a response and all it's parts that
can be converted to json

Changes:
- Create prompt.py
- Formalize craft response dataclass
- Create class that can be serialized
- looking at tests, but we need to import PyQt5 parts directly not through krita.
- allow for text in front of the json
- from_history_json()
- Add prompt_from_selection
- Remove all base64 from the response path
- All responses are now PromptResponse
- Assert images doesn't exceed names

JasonS09 commented:
> * Is this a reasonable design
> * Is it worth making an "info" class for holding generation information but no the base64 encoded images. -- this allows to/from json
>   -- central list of properties

I think we can go with the class approach. May be a simpler and less confusing design than having to use the `PROMPT_LOADER_KEYS` list.


> * Should we update the model I don't because the user can't see it from the prompt screen. (expand info bar maybe?)

One possible way to update the model could be adding this in `update_prompt_params` function:

```
dockers = Krita.instance().dockers()
for d in dockers:
    if d.objectName() == TAB_SDCOMMON:
    d.page_widget.sd_model_layout.qcombo.setEditText(params[keys[0]])
    break
```

ElliottLester commented:
@JasonS09 
So I wrote a replacement for craft_response() that formalizes the response class and pushes some code out of client.py
Let me know if you think this is the right direction?
There are some TODOs in this PR I would like your thoughts on.
client.py:224
prompt.py

Other Querys?
- Are we able to import PyQT directly on other platforms? this makes writing unit tests easier.
- These classes can be printed and serialized to json
- The addition of  Base64Image class lets us move the QImage and b64string around together this should stop us from changing formats everywhere.
- I would like to move the workflow generation code out of client.py I'm thinking Prompt.to_workflow()

JasonS09 commented on frontends/krita/krita_comfy/client.py:
```diff
@@ -257,10 +221,11 @@ def on_image_received(img):
                 assert img is not None, "Backend Error, check terminal"
                 qimage = QImage()
                 qimage.loadFromData(img)
+                # TODO pass QImage directly
```
The reason why we encoded to base64 here is because script's `img_inserter` expects encoded image data. I think you're already handling that in `PromptResponse` class, so it should be good if we just pass the raw data directly.

JasonS09 commented:
> * I would like to move the workflow generation code out of client.py I'm thinking Prompt.to_workflow()

Go for it.

ElliottLester commented:
I have removed all the base64 conversions in the response section.
I think I'll check this in before moving onto the Prompt/config side.

ElliottLester committed:
Fix ControlNet Preview

ElliottLester committed:
Basic Unit-test for PromptResponse

ElliottLester committed:
Save history to PromptResponse

for debugging

ElliottLester committed:
Recover from empty response:

We should check if the response has images in it before we try to insert them.
Then notify the user the Server sent us an empty response.
Added:
ERR_BACKEND
ERR_EMPTY_RESPONSE

ElliottLester committed:
Add support for SSL
